### PR TITLE
Add a way to use the Repm HTTP server

### PR DESCRIPTION
**Pull request by arv**
This is usefule because gomobile takes a long time to rebuild and using
an http server allows faster iterations.

### CODE REVIEW

**Comment by arv** [comment]
It is not clear to me how to best expose this. My current plan is to  make `Replicache` constructor and the static functions (`list` and `drop`) take an optional invoke function or invoker instance.

I will do that tomorrow unless you have any cleaner suggestions?

**Comment by aboodman** [comment]
How about having a static string property `repmURL`. If `null`, `repm` is in-process. Otherwise, it is out of process, contacted via HTTP at the specified URL. The actual branch can be done via a special invoker fn as you describe, or just a branch.

**Comment by arv** [comment]
I settled on an `repmInvoke` parameter. Ended up being useful for the tests too.

File: lib/replicache.dart [hunk 1]
```diff
@@ -75,40 +75,37 @@ class Replicache implements ReadTransaction {
   bool _closed = false;
   Set<_Subscription> _subscriptions = Set();
   Future<void> _syncFuture;
+  RepmInvoke _repmInvoke;
 
   /// Lists information about available local databases.
-  static Future<List<DatabaseInfo>> list() async {
-    var res = await _staticInvoke('', 'list');
+  static Future<List<DatabaseInfo>> list({RepmInvoke repmInvoke}) async {
```
Comment (aboodman) [inline lib/replicache.dart:81]: Oh yeah, this seems reasonable given the current design.

File: lib/replicache.dart [hunk 2]
```diff
@@ -135,32 +134,33 @@ class Replicache implements ReadTransaction {
...
   }
 
   /// Sets the level of verbosity Replicache should log at.
   static setLogLevel(LogLevel level) {
     globalLogLevel = level;
-    _staticInvoke(
-        '',
-        'setLogLevel',
-        {
+    _staticInvoke('', 'setLogLevel',
```
Comment (aboodman) [inline lib/replicache.dart:154]: I landed the change to make this a getter/setter so this needs to be rebased. Also, doesn't this require the invoke param?

File: lib/replicache.dart [hunk 3]
```diff
@@ -256,14 +256,17 @@ class Replicache implements ReadTransaction {
 
   Future<void> _sync() async {
     final syncHead = await _beginSync();
-    await _maybeEndSync(syncHead);
+    if (syncHead != '00000000000000000000000000000000') {
```
Comment (aboodman) [inline lib/replicache.dart:259]: It seems like maybe you are testing against HEAD of `repm`. Now that the version pinning is landed, you should switch to using it.

File: lib/replicache.dart [hunk 4]
```diff
@@ -715,3 +712,61 @@ class _MutateResult<R> {
   final String ref;
   _MutateResult(this.result, this.ref);
 }
+
+typedef Future<dynamic> RepmInvoke(String dbName, String rpc, [dynamic args]);
```
[the remark this answers is not in the document]
Comment (arv) [reply lib/replicache.dart:719]: Yes. Time to break it up.

[the remark this answers is not in the document]
Comment (arv) [reply lib/replicache.dart]: I'm going to do more code splitting in a follow up.

File: lib/replicache.dart (hunk 2 above)
Comment (arv) [reply lib/replicache.dart:176]: setters cannot have extra params `x = y` and all that.

We would need to make this a method too but I'm not sure I care enough. We can put it on the ReplicacheTest instance... Which by the way I'm going to move to lib/src/ in a follow up

**Comment by arv** [comment]
TBR

**Pull request merged**